cmd/main.go

```diff
@@ -2,16 +2,188 @@ package main
+	"github.com/datawire/ambassador-agent/pkg/agent"
+	"github.com/google/uuid"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/klog/v2"
+
+	"github.com/datawire/dlib/dgroup"
+	"github.com/datawire/dlib/dlog"
+	"github.com/emissary-ingress/emissary/v3/pkg/busy"
+	"github.com/emissary-ingress/emissary/v3/pkg/logutil"
+)
+
+// internal k8s service
+const (
+	AdminDiagnosticsPort     = 8877
+	DefaultSnapshotURLFmt    = "http://ambassador-admin:%d/snapshot-external"
+	DefaultDiagnosticsURLFmt = "http://ambassador-admin:%d/ambassador/v0/diag/?json=true"
+
+	ExternalSnapshotPort = 8005
+
+	leaseLockName = "ambassador-agent-lease-lock"
 )
 
 func main() {
-	err := agent.Main(context.Background(), "", os.Args[1:]...)
+	ctx := context.Background()
+
+	// creates the in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		klog.Fatal(err.Error())
+	}
+	// creates the clientset
+	clientset := kubernetes.NewForConfigOrDie(config)
+
+	agentNamespace := getEnvWithDefault("AGENT_NAMESPACE", "ambassador")
+
+	ambAgent := agent.NewAgent(nil, agent.NewArgoRolloutsGetter, agent.NewSecretsGetter, clientset, agentNamespace)
+
+	// all log things need to happen here because we still allow the agent to run in amb-sidecar
+	// and amb-sidecar should control all the logging if it's kicking off the agent.
+	// this codepath is only hit when the agent is running on its own
+	logLevel := os.Getenv("AES_LOG_LEVEL")
+	// by default, suppress everything except fatal things
+	// the watcher in the agent will spit out a lot of errors because we don't give it rbac to
+	// list secrets initially.
+	klogLevel := 3
+	if logLevel != "" {
+		logrusLevel, err := logutil.ParseLogLevel(logLevel)
+		if err != nil {
+			dlog.Errorf(ctx, "error parsing log level, running with default level: %+v", err)
+		} else {
+			busy.SetLogLevel(logrusLevel)
+		}
+		klogLevel = logutil.LogrusToKLogLevel(logrusLevel)
+	}
+	klogFlags := flag.NewFlagSet(os.Args[0], flag.ContinueOnError)
+	klog.InitFlags(klogFlags)
+	if err := klogFlags.Parse([]string{fmt.Sprintf("-stderrthreshold=%d", klogLevel), "-v=2", "-logtostderr=false"}); err != nil {
+		klog.Fatal(err.Error())
+	}
+	snapshotURL := getEnvWithDefault("AES_SNAPSHOT_URL", fmt.Sprintf(DefaultSnapshotURLFmt, ExternalSnapshotPort))
+	diagnosticsURL := getEnvWithDefault("AES_DIAGNOSTICS_URL", fmt.Sprintf(DefaultDiagnosticsURLFmt, AdminDiagnosticsPort))
+
+	reportDiagnostics := os.Getenv("AES_REPORT_DIAGNOSTICS_TO_CLOUD")
+	if reportDiagnostics == "true" {
+		ambAgent.SetReportDiagnosticsAllowed(true)
+	}
+
+	metricsListener, err := net.Listen("tcp", ":8080")
 	if err != nil {
-		fmt.Fprintf(os.Stdout, "error executing from argparser: %v", err)
-		os.Exit(1)
+		klog.Fatal(err.Error())
+	}
+	dlog.Info(ctx, "metrics service listening on :8080")
```

Comment:
Are we using both `dlog` and `klog`?
If so, when would you use one over the other?

Reply:
just dlog probs